### PR TITLE
[16.0][MIG] project_task_code

### PR DESCRIPTION
**Pull request by Abranes**
Supersedes #1013

### CODE REVIEW

**Comment by rafaelbn** [comment]
Hello @Abranes 

Please this module in 15 (at least) is creating constraint that breaks runboat with other modules. could you FIX it here in migration?

@Shide  
> Yes, project_task_code module is creating a constraint on project.task and other modules who creates tasks (by strange ways or on a different order) breaks the init of the ruboat with demo data.

@sbidoul 
> Ah, yes. IMO modules that add constraints on standard models should enable them by configuration, otherwise we reach dead ends like this one.

https://github.com/OCA/project/pull/1096#issuecomment-1484977104

Thank you!

**Comment by Abranes** [comment]
> Hello @Abranes
> 
> Please this module in 15 (at least) is creating constraint that breaks runboat with other modules. could you FIX it here in migration?
> 
> @Shide
> 
> > Yes, project_task_code module is creating a constraint on project.task and other modules who creates tasks (by strange ways or on a different order) breaks the init of the ruboat with demo data.
> 
> @sbidoul
> 
> > Ah, yes. IMO modules that add constraints on standard models should enable them by configuration, otherwise we reach dead ends like this one.
> 
> [#1096 (comment)](https://github.com/OCA/project/pull/1096#issuecomment-1484977104)
> 
> Thank you!

Done! You can review it.

File: project_task_code/hooks.py
Comment (Shide) [inline project_task_code/hooks.py]: Should be done only when the task code is activated? It's deactivated by default.
Same on `post_init_hook`

File: project_task_code/models/project_task.py
Comment (Shide) [inline project_task_code/models/project_task.py:39]: Should be done if user has activated the corresponding parameter in `res.config.settings` and sequence exists in `self.env.company`.

File: project_task_code/models/res_config_settings.py
Comment (Shide) [inline project_task_code/models/res_config_settings.py]: I think could be a `ir.sequence` record (for multi company purposes)

File: project_task_code/models/project_task.py
Comment (Shide) [inline project_task_code/models/project_task.py]: No need to check for parameter. This should pass even if the parameter is not set. Only check it if task has code.

**Review by Shide — CHANGES_REQUESTED** [review]
Code is being set even if in the settings, the Task Code Unique is not checked.
Also check the `pre_init_hook` and `post_init_hook`

**Comment by rousseldenis** [comment]
@Abranes

File: project_task_code/views/project_view.xml
Comment (jort-aardug) [inline project_task_code/views/project_view.xml]: Field "name" exists twice in the kanban view, once for non-active tasks (shows with strikethrough), once for active tasks. This field selector only targets the first one, so you only see the code field on archived tasks.

**Comment by yaazkal** [comment]
Is there something pending here still?

**Comment by Shide** [comment]
> Is there something pending here still?

Check @jort-aardug and @Shide suggestions

File: project_task_code/models/project_task.py
Comment (robinkeunen) [reply project_task_code/models/project_task.py]: I think there is a confusion here in the use of the parameter : 

- the task code is generated when the module is installed
- the parameter adds the check for unicity

**Review by robinkeunen — APPROVED** [review]

**Comment by yaazkal** [comment]
@Abranes can you check the suggestions for this to be merged? Thanks

File: project_task_code/views/project_view.xml
Comment (robinkeunen) [reply project_task_code/views/project_view.xml]: I tried fixing it but it does it actually does not interact well with the `name_with_subtask_count` widget : 

- there is no space between code and name fields
- only the code in stroke trough when the task is archived.

Screenshot 2023-10-26 at 16 38 52
Screenshot 2023-10-26 at 16 39 17

Comment (robinkeunen) [reply project_task_code/views/project_view.xml]: This
```xml
<xpath expr="//field[@name='name']/.." position="before">
    <field name="code" class="me-2"/>
</xpath>
```
produces

Screenshot 2023-10-26 at 17 17 39


For both archived and unarchived tasks. The `s` tag is overiden by the `name_with_subtask_count` widget.

**Comment by robinkeunen** [comment]
I made a suggestion for the _name tag_ comment. If I understand correctly, the `project_task_unique_code` parameter adds the check for unicity, it does not activate the feature. If so, @Shide 's comment do not apply.

I can do a superseeding PR if @Abranes or maintainers give me the go.

**Comment by Abranes** [comment]
Sorry guys, I forgot this PR.

I'll try to address all points next week.

File: project_task_code/hooks.py
Comment (Abranes) [reply project_task_code/hooks.py]: Removed, not needed because it's deactivated by default.

File: project_task_code/models/res_config_settings.py
Comment (Abranes) [reply project_task_code/models/res_config_settings.py]: Done

File: project_task_code/models/project_task.py
Comment (Abranes) [reply project_task_code/models/project_task.py]: Changed because now the code is only set if the sequence is set in configuration, so now the constrains check if the task has code and if any other task in the company has the same code.

File: project_task_code/views/project_view.xml
Comment (Abranes) [reply project_task_code/views/project_view.xml]: Done

**Comment by Abranes** [comment]
@jort-aardug @Shide @robinkeunen @yaazkal could you check it one more time? :)

**Review by Shide — APPROVED** [review]
**praise:** Good job with this man :)

Code LGTM

**Review by robinkeunen — CHANGES_REQUESTED** [review]
I think there is a confusion here, the module should

- add the task code upon installation
- force the unicity constraint on configuration

The last fixes does the opposite:

- force the unicity constraint upon installation
- create the task code on configuration

**Comment by Abranes** [comment]
> I think there is a confusion here, the module should
> 
> * add the task code upon installation
> * force the unicity constraint on configuration
> 
> The last fixes does the opposite:
> 
> * force the unicity constraint upon installation
> * create the task code on configuration

It's what @Shide requests and I think it's not a bad approach.

**Comment by robinkeunen** [comment]
But that changes the purpose of the module, migration is not the good time to do that. Plus a better module name would then me `project_task_code_unique`.

Plus it does not follow @sbidoul and @rafaelbn 's recommendation. Let's see what others think.

**Comment by robinkeunen** [comment]
Hi @rousseldenis  @dreispt , we're disagreeing on the way to go with this PR. Can you help us move forward ?

**Comment by Shide** [comment]
> But that changes the purpose of the module, migration is not the good time to do that. Plus a better module name would then me `project_task_code_unique`.
> 
> Plus it does not follow @sbidoul and @rafaelbn 's recommendation. Let's see what others think.

Maybe we should add a migration script?
 
The module without this fix/feature will break tests for sure at any point in the future... If someone has the bad luck of having this module installed in another way.

**Comment by rafaelbn** [comment]
Hello,

I don't know in this point which is the disagree. I pinged @Abranes because this module is breaking other modules test, etc... and we said to FIX it here.

> Hello @Abranes
> 
> Please this module in 15 (at least) is creating constraint that breaks runboat with other modules. could you FIX it here in migration?
> 
> @Shide
> 
>     Yes, project_task_code module is creating a constraint on project.task and other modules who creates tasks (by strange ways or on a different order) breaks the init of the ruboat with demo data.
> 
> @sbidoul
> 
>     Ah, yes. IMO modules that add constraints on standard models should enable them by configuration, otherwise we reach dead ends like this one.
> 
> https://github.com/OCA/project/pull/1096#issuecomment-1484977104
> 
> Thank you!

- [ ] The request was to address a constraint issue in the project_task_code module for version 15. The module's constraint on project.task is causing conflicts with other modules in runboat environments, especially those creating tasks in unconventional ways. @sbidoul suggests that modules adding constraints to standard models should have the option to enable these constraints through configuration to avoid such conflicts. This suggestion aims to prevent dead ends and improve module compatibility.

We are not using this module at all, just we are affected in other modules un runboat and test because this one.

@robinkeunen > But that changes the purpose of the module
@Shide > The module without this fix/feature will break tests for sure at any point in the future... If someone has the bad luck of having this module installed in another way.

- The purpose of the module shouln't be change but FIXED

Maybe @oihane or @pedrobaeza as original contributors what to put an opinion. I don't know exacly what's the problem now. @dreispt Do you have a solution here o some past use case?

Thank you all! 😄 ❤️

**Comment by pedrobaeza** [comment]
Can you summarize instead of copying a lot of texts? If I understand correctly, there was a constraint and it has been removed in the migration? Or the contrary? Which is the problem in runboat?

**Comment by robinkeunen** [comment]
In 15.0 : `project_task_code`

- adds `code` field on tasks
- adds sql contraint `project_task_unique_code`
- a hook initializes code from `project.task` sequence
- in `create`, code is assigned from `project.task` sequence

In his migration review (2023-03-27), @rafaelbn pointed out that the a configuration should be added for the constraint. Otherwise, it breaks the tests from other modules.

I understand from the suggestion that the module should

- add the task code upon installation
- force the unicity constraint on configuration

The fix from @abranes, based on @shide's request, does the opposite, it 

- forces the unicity constraint upon installation
- creates the task code on configuration

**Comment by pedrobaeza** [comment]
Thanks for the summary, Robin.

It's true that enforcing a uniqueness can be conflicting for prior version databases not containing this constraint migrating to this version. If implemented via SQL constraint, Postgres will try to put the constraint and fail, not having the constraint at all, but only a warning each time you update. Doing that by Python constraint allows to be enabled by configuration, and this would be a saner feature.

Another conflicting change is to fill existing taxes with a code, as you may not want to have one at all. I would go for having a button on configuration to fill the code on "empty code" taxes on demand.

Do you see it very complicated to be done?

**Comment by huguesdk** [comment]
i don’t understand why so many changes are introduced in a simple version migration. this sql constraint is present since version 8. the hooks that run upon module initialization ensure that this constraint is satisfied, and that tasks created subsequently (even by other modules) will have a unique code (coming from a sequence). it shouldn’t cause an issue.

i don’t think it’s a good idea to introduce such significant changes as part of a version migration, unless something that changed in that version prevents the module from working as it did in the previous version.

i think that the changes in commit https://github.com/OCA/project/commit/90103ce1cdf1520131997f6e6de11396cf2351ac are enough.

**Comment by robinkeunen** [comment]
Rather than arguing, I propose this PR #1209. I leave it to the PSC to choose what is more appropriate :-)

**Comment by huguesdk** [comment]
could someone give a more detailed explanation about what the problem that we’re trying to solve exactly is (apart from porting this module to 16)? i’ve read that it caused errors during the runboat build. could someone copy the exact error here to see if it can be solved in a simpler way?

note that the original sql constraint is not multi-company compatible (it requires uniqueness of the code while it should probably be unique per company) while sequences are per company by default. maybe this could be the issue?

**Comment by legalsylvain** [comment]
Hi @rafaelbn 

> Please this module in 15 (at least) is creating constraint that breaks runboat with other modules. could you FIX it here in migration?

could you mention the error generated ? I don't understand how it can break things !

Thanks !

**Comment by Abranes** [comment]
Hi all! I can make the changes you want but first let's all be clear, because in the first instance I perform a normal migration, then modify the module so that the constraint was not by SQL and was by python and thus be able to activate or deactivate it in the configuration to avoid problems with other modules, and finally carry out the current implementation, all at the request of different people.

How do you prefer it to proceed?

**Comment by huguesdk** [comment]
i personally think that the sql constraint should stay. it is the only way to ensure uniqueness in a reliable way. checking it in python won’t prevent the same code to be used if records are created at the same time. we need to check what problem the sql constraint actually caused, and fix it without removing the constraint.

**Comment by rafaelbn** [comment]
Thinking in an OCA Repo with more modules involved:

IMO modules that add constraints on standard models should enable them by configuration, otherwise we reach dead ends. So not by SQL but by python and thus be able to activate or deactivate it in the configuration to avoid problems with other modules

Not blocking anything, just my opinion.

Thanks all by efforts in the best way! 😄 ❤️

**Comment by legalsylvain** [comment]
> IMO modules that add constraints on standard models should enable them by configuration, otherwise we reach dead ends. So not by SQL but by python and thus be able to activate or deactivate it in the configuration to avoid problems with other modules

Hi. my opinion is "Keep It simple". An SQL constraint is way more simple as constraint enabled by configuration. So, if there is no existing problem, I propose to keep the V15 code. 

https://github.com/OCA/project/blob/15.0/project_task_code/models/project_task.py#L18-L20 is simpler (**2 lines**) than : 
- https://github.com/OCA/project/pull/1084/files#diff-c6bf6da19dd508004fa0507f6c93818e4d02c9cc2ccf5a8ef3b86482959e59d2R37-R63 (**37 lines**)
- https://github.com/OCA/project/pull/1084/files#diff-5cd651cf3b3f58da9ee3f1c78edea6215d0464137e4e03300db36cddb4edf6faR1-R26 (**26 lines**)
- https://github.com/OCA/project/pull/1084/files#diff-b66987146c83cf675c185c5bf63cdf252472583bd56acd42a61bc8701c5b26afR1-R14 (**14 lines**).

@rafaelbn : you was saying that V15 runboat was broken, but I see it green. Is it fixed ? https://runboat.odoo-community.org/webui/builds.html?repo=OCA/project&target_branch=15.0

**Comment by rafaelbn** [comment]
Hello @legalsylvain! Happy new year 2024! 😄 

@Shide will explain after holidays the problem that this modules causes when living in the same repo with other ones. 

We are not using this module at all, we just found problems when working in others in the same repo.

Thank you!

**Comment by Shide** [comment]
@legalsylvain the case which i'm saying this needs a change is because this happens to me in this PR: https://github.com/OCA/project/pull/1096#issuecomment-1484977104

Check out the next 4 comments where i'd invoked you because the runboat breaks

**Comment by legalsylvain** [comment]
Hi @Shide. Thanks a lot for pointing the issue. @robinkeunen, could you take a look at this issue ? 
thanks ! 
and happy new years !

**Comment by rafaelbn** [comment]
This PR is continued in https://github.com/OCA/project/pull/1209

**Pull request closed without merging**
